chia/data_layer/data_layer.py

```diff
@@ -224,105 +183,92 @@ async def fetch_and_validate(self, subscription: Subscription) -> None:
-                await self.data_store.rollback_to_generation(tree_id, 0)
-                raise RuntimeError("Could not validate on-chain data. Downloading from scratch as a fallback.")
-
-        self.log.info(
-            f"Finished downloading and validating {subscription.tree_id}. "
-            f"Wallet generation saved: {singleton_record.generation}. "
-            f"Root hash saved: {singleton_record.root}."
-        )
-        await self.data_store.set_validated_wallet_generation(tree_id, int(singleton_record.generation))
-
-    async def subscribe(self, store_id: bytes32, mode: DownloadMode, ip: str, port: uint16) -> None:
-        subscription = Subscription(store_id, mode, ip, port)
+                self.log.warning(f"Exception while downloading files for {tree_id}: {e} {traceback.format_exc()}.")
+
+    async def upload_files(self, tree_id: bytes32) -> None:
+        singleton_record: Optional[SingletonRecord] = await self.wallet_rpc.dl_latest_singleton(tree_id, True)
+        if singleton_record is None:
+            self.log.info(f"Upload files: no on-chain record for {tree_id}.")
+            return
+        root = await self.data_store.get_tree_root(tree_id=tree_id)
+        publish_generation = min(singleton_record.generation, 0 if root is None else root.generation)
+        # If we make some batch updates, which get confirmed to the chain, we need to create the files.
+        # We iterate back and write the missing files, until we find the files already written.
+        root = await self.data_store.get_tree_root(tree_id=tree_id, generation=publish_generation)
+        while publish_generation > 0 and await write_files_for_root(
+            self.data_store,
+            tree_id,
+            root,
+            self.server_files_location,
+        ):
+            publish_generation -= 1
+            root = await self.data_store.get_tree_root(tree_id=tree_id, generation=publish_generation)
+
+    async def add_missing_files(self, store_id: bytes32, override: bool, foldername: Optional[Path]) -> None:
+        root = await self.data_store.get_tree_root(tree_id=store_id)
+        singleton_record: Optional[SingletonRecord] = await self.wallet_rpc.dl_latest_singleton(store_id, True)
+        if singleton_record is None:
+            self.log.error(f"No singleton record found for: {store_id}")
+            return
+        max_generation = min(singleton_record.generation, 0 if root is None else root.generation)
+        server_files_location = foldername if foldername is not None else self.server_files_location
+        for generation in range(1, max_generation + 1):
+            root = await self.data_store.get_tree_root(tree_id=store_id, generation=generation)
+            await write_files_for_root(self.data_store, store_id, root, server_files_location, override)
+
+    async def subscribe(self, store_id: bytes32, urls: List[str]) -> None:
+        parsed_urls = [url[:-1] if url.endswith("/") else url for url in urls]
```

Comment:
Will remove any trailing `/`s.

```suggestion
        parsed_urls = [url.rstrip("/") for url in urls]
```